// chrome-devtools-mcp/build/src/tools/categories.js
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
export var ToolCategory;
(function (ToolCategory) {
    ToolCategory["INPUT"] = "input";
    ToolCategory["NAVIGATION"] = "navigation";
    ToolCategory["EMULATION"] = "emulation";
    ToolCategory["PERFORMANCE"] = "performance";
    ToolCategory["NETWORK"] = "network";
    ToolCategory["DEBUGGING"] = "debugging";
    ToolCategory["EXTENSIONS"] = "extensions";
    ToolCategory["THIRD_PARTY"] = "experimentalThirdParty";
    ToolCategory["MEMORY"] = "memory";
    ToolCategory["WEBMCP"] = "experimentalWebmcp";
})(ToolCategory || (ToolCategory = {}));
export const labels = {
    [ToolCategory.INPUT]: 'Input automation',
    [ToolCategory.NAVIGATION]: 'Navigation automation',
    [ToolCategory.EMULATION]: 'Emulation',
    [ToolCategory.PERFORMANCE]: 'Performance',
    [ToolCategory.NETWORK]: 'Network',
    [ToolCategory.DEBUGGING]: 'Debugging',
    [ToolCategory.EXTENSIONS]: 'Extensions',
    [ToolCategory.THIRD_PARTY]: 'Third-party',
    [ToolCategory.MEMORY]: 'Memory',
    [ToolCategory.WEBMCP]: 'WebMCP',
};
export const OFF_BY_DEFAULT_CATEGORIES = [
    ToolCategory.EXTENSIONS,
    ToolCategory.THIRD_PARTY,
    ToolCategory.WEBMCP,
];
